### PR TITLE
Support `batchbald` (BatchBALD) as a certainty metric

### PR DESCRIPTION
**Pull request by Leengit**
Add `batchbald` to the list of supported certainty metrics, `confidence`, `margin`, and `negative_entropy`.  This pull request extends work begun in Pull Request #12 to a minimum viable product that works on basic cases.  (Future work can find and fix edge cases, possibly tweak high-level parameters so as to train with fewer labels, and possibly address any performance issues so as to run faster per label.)

This pull request also raises the Python version to 3.11 and restores the optional `torch` Python package dependency.

This code works on my computer, but ... reviewers / testing should test the new `batchbald` functionality to verify that it also works for you.  Also please test one or more of the previous certainty metrics to make sure that that code, which is tensorflow-based, is not broken by this new code that is torch-based.

### CODE REVIEW

**Comment by Leengit** [comment]
They are not sorted after the `batchbald_batch_size` is exhausted.  They are all tied at that point, so it all depends upon how the sort algorithm in the guided labeling section breaks ties.  (For example if it is a stable sort ....)  As I see it, possible behaviors at that point include: (A) refuse to give more superpixels (and prompt the user to retrain), (B) do whatever is necessary on the sly as the first batch is near exhausted, so as to simulate having an arbitrarily long sorted list of superpixels (C) fall back to some other criterion, or (D) arbitrarily.  The latter two represent (somewhat of) a lie to the user who requested batchbald.  The extent to which C would be stronger than D for active learning is not clear to me; perhaps neither is particularly useful after the first 100 are labeled.  If the sort in the guided labeling is a stable sort, a possible implementation of C is to first sort all superpixels according to the fallback criterion and then sort them all according to the batchbald criterion.

Perhaps this is a question for Jeff G. and Lee C.

Yes, if you'd be willing to write down your specific ideas for progress reporting and testing, as GitHub issues, I'd appreciate that.

**Review by manthey — APPROVED** [review]
Let's discuss the fallback choices at the next meeting.  This has worked for me in several instances.  I say we merge it and if there are any undiscovered issues we can deal with them then.
